clang/lib/Analysis/FlowSensitive/Models/ReverseNullModel.cpp

```diff
@@ -0,0 +1,723 @@
+    assert(Var != nullptr);
+
+    auto *RootValue = getValue(*Var, Env);
+    if (!RootValue) {
+       return;
+    }
+
+    auto *NewRootValue = Env.createValue(Var->getType());
+    if (!NewRootValue) 
+        return;
+
+    setLValue(*Var, *NewRootValue, Env);
+
+    Arena &A = Env.arena();
+
+    if (IsNonnullBranch) {
+        BoolValue &IsNonnull = getVal(kIsNonnull, *RootValue);
+        BoolValue &IsNull = getVal(kIsNull, *RootValue);
+
+        Env.addToFlowCondition(A.makeNot(IsNull.formula()));
+        NewRootValue->setProperty(kIsNull, Env.getBoolLiteralValue(false));
+
+        NewRootValue->setProperty(kIsNonnull, IsNonnull);
+    } else {
+        BoolValue &IsNonnull = getVal(kIsNonnull, *RootValue);
+        BoolValue &IsNull = getVal(kIsNull, *RootValue);
+
+        NewRootValue->setProperty(kIsNull, IsNull);
+
+        Env.addToFlowCondition(A.makeNot(IsNonnull.formula()));
+        NewRootValue->setProperty(kIsNonnull, Env.getBoolLiteralValue(false));
+    }
+}
+
+void matchNullptrExpr(const Expr *expr, const MatchFinder::MatchResult &Result, Environment &Env) {
+    const auto *PrVar = Result.Nodes.getNodeAs<Expr>(kVar);
+    assert(PrVar != nullptr);
+
+    if (Env.getValue(*PrVar))
+      return;
+
+    // FIXME: Nullptr does not create a value
+    auto *RootValue = Env.createValue(PrVar->getType());
+    if (!RootValue) {
+       return;
+    }
+
+    RootValue->setProperty(kIsNull, Env.getBoolLiteralValue(true));
+    RootValue->setProperty(kIsNonnull, Env.getBoolLiteralValue(false));
+    Env.setValue(*PrVar, *RootValue);
+}
+
+void matchAddressofExpr(const Expr *expr, const MatchFinder::MatchResult &Result, Environment &Env) {
+    const auto *PrVar = Result.Nodes.getNodeAs<Expr>(kVar);
+    assert(PrVar != nullptr);
+
+    auto *RootValue = Env.createValue(PrVar->getType());
+    if (!RootValue) {
+       return;
+    }
+
+    RootValue->setProperty(kIsNull, Env.getBoolLiteralValue(false));
+    RootValue->setProperty(kIsNonnull, Env.getBoolLiteralValue(true));
+    Env.setValue(*PrVar, *RootValue);
+}
+
+void matchFunctionCallExpr(const Expr *fncall, const MatchFinder::MatchResult &Result, Environment &Env) {
+    // This is not necessarily a prvalue, since operators such as prefix ++ are lvalues.
+    const auto *Var = Result.Nodes.getNodeAs<Expr>(kVar);
+    assert(Var != nullptr);
+
+    // Initialize to (Unknown, Unknown)
+    if (Env.getValue(*Var))
+      return;
+
+    auto *RootValue = Env.createValue(Var->getType());
+    if (!RootValue)
+       return;
+    
+    initializeRootValue(*RootValue, Env);
+    setUnknownValue(*Var, *RootValue, Env);
+}
+
+ReverseNullDiagnoser::ResultType
+diagnoseCastExpr(const CastExpr *Stmt, const MatchFinder::MatchResult &Result,
+               const Environment &Env) {
+    const auto *Var = Result.Nodes.getNodeAs<Expr>(kVar);
+    assert(Var != nullptr);
+
+    Arena &A = Env.arena();
+
+    if (auto *RootValue = Env.getValue(*Var)) {
+        // FIXME: Hack
```

Comment:
What is the hack here? What happens if the hack is removed, or how should it disappear?